### ISSUE TITLE
Sprite: Raycasting with offscreen sprites fails

### ISSUE THREAD

**Issue by Mugen87** [mrdoob/three.js#14624]
##### Description of the problem

A user in the forum reported a bug in `Sprite.raycast()`.

https://discourse.threejs.org/t/problematic-raycasting-of-offscreen-three-sprite-object/3644

The method depends on `Object3D.modelViewMatrix`. But this matrix is only updated if the sprite is actually rendered. If not, `modelViewMatrix` does not regard the actual object's transformation. This means that `Sprite.raycast()` detects intersections of sprites outside the view frustum.

https://github.com/mrdoob/three.js/blob/1ffaa41a4b5f053ad62da1682844b53b041e63ca/src/objects/Sprite.js#L99

##### Three.js version

- [x] Dev
- [x] r95
- [ ] ...

##### Browser

- [x] All of them
- [ ] Chrome
- [ ] Firefox
- [ ] Internet Explorer

##### OS

- [ ] All of them
- [ ] Windows
- [ ] macOS
- [ ] Linux
- [ ] Android
- [ ] iOS

**Comment by Mugen87** [comment]
Test case: https://jsfiddle.net/f2Lommf5/10724/

**Comment by 06wj** [comment]
The problem is how to get the reference of the camera to calculate ```modelViewMatrix```.

**Comment by Mugen87** [comment]
One solution is to calculate `modelViewMatrix` for sprites in `projectObject()`.

**Comment by Mugen87** [comment]
Another user ran into this bug: https://discourse.threejs.org/t/cannot-get-the-object-of-the-click/7451

Unfortunately, I can't think of another "easy" fix than proposed in #14679.

**Comment by WestLangley** [comment]
I would not consider this a bug if user is frustum-culling the sprite.

The user should set:
```js
sprite.frustumCulled = false;
```
if the user expects raycasting to work on a sprite outside of the frustum.

**Comment by Sayid1** [comment]
> I would not consider this a bug if user is frustum-culling the sprite.
> 
> The user should set:
> 
> ```js
> sprite.frustumCulled = false;
> ```
> 
> if the user expects raycasting to work on a sprite outside of the frustum.

How do you konw a sprite at outside of the frustum?

**Comment by Mugen87** [comment]
> I would not consider this a bug if user is frustum-culling the sprite.

I don't think so. The fiddle from the forum clearly demonstrates a bug since `Raycaster.intersectObjects()` returns intersections although the ray visually misses all objects. The problem is that sprites outside the view frustum have an initial `modelViewMatrix` which lead to the wrong result.

**Comment by WestLangley** [comment]
Thanks, @Mugen87. I agree with you that generating false positives is not acceptable.

Is it true that sprite raycasting works correctly when `sprite.frustumCulled = false;`?

**Comment by Mugen87** [comment]
Yes, that is correct.

**Comment by WestLangley** [comment]
Raycasting does not require a camera, obviously, but raycasting against a sprite only makes sense if it is with respect to a camera. So we have a design limitation.

The only solutions I can think of seem hacky to me, but then again, what is hacky is subjective.

Setting `sprite.frustumCulled = false;` is a work-around, but I agree with @Mugen87 that is not very appealing.

**Comment by Mugen87** [comment]
> The only solutions I can think of seem hacky to me, but then again, what is hacky is subjective.

Can you please explain your solution? I'm afraid this issue will frequently pop up and I do prefer a fix in the library. #14679 would definitely work.

**Comment by WestLangley** [comment]
>The only solutions I can think of seem hacky to me
>>Can you please explain your solutions?

1. Pass camera to raycast.

2. Add `sprite.supportRaycast` flag.

3. #14679 - not really appealing to me - most sprites do not require raycast support.

Also, this `Sprite.raycast()` code is computing `camera.matrixWorld`.
```js
viewWorldMatrix.getInverse( this.modelViewMatrix ).premultiply( this.matrixWorld );
```
It would be nice if that code could be replaced.

**Comment by Mugen87** [comment]
I vote for option 1 then.

**Comment by 06wj** [comment]
I vote for 1 too.It will also fix #14936

**Comment by WestLangley** [comment]
@06wj Do you want to propose a PR, since you have been working on this issue?

**Comment by WestLangley** [comment]
Oh man, this is really going to be ugly IMO, but I won't fight it because I don't have a better solution...

Maybe option 3 is better after all...

**Comment by Mugen87** [comment]
Unfortunately, option 3 does not fix #14936. Sooner or later this issue will also be reported by users.

@06wj A PR with option 1 would be great 😊

**Comment by 06wj** [comment]
https://github.com/mrdoob/three.js/blob/93e72ba7b24958ddb0652bd33171edd14ed2d693/src/core/Raycaster.js#L96

Should ```raycaster.intersectObject``` add```camera``` param?

**Comment by WestLangley** [comment]
>Should raycaster.intersectObject add camera param?

I don't think so. I experimented with #16423 and it seems to work -- but it was just a quick test...

**Comment by WestLangley** [comment]
Closing. Should be fixed in #16423.
